File: akka-samples/scala-samples/src/main/scala/actors/watch/WatchPrinter.scala
package actors.watch

import akka.actor.typed.scaladsl.Behaviors
import akka.actor.typed.{ActorSystem, Behavior, Terminated}

import scala.util.Random

object WatchPrinter {
  def main(args: Array[String]): Unit = {
    ActorSystem[String](Printer(), "watch-printer")
  }
}

private object Printer {
  def worker(): Behavior[String] =
    Behaviors.setup[String] { context =>
      Thread.sleep(Random.between(2000, 10000))
      Behaviors.stopped
    }

  def apply(): Behavior[String] =
    Behaviors.setup[String] { context =>

      context.watch(context.spawnAnonymous(worker()))
      context.watchWith(context.spawnAnonymous(worker()), "Worker finished!")

      Behaviors.receive[String] { (_, message) =>
        println(message)

        Behaviors.same
      }.receiveSignal {
        case (_, Terminated(ref)) =>
          println(s"Actor '${ref.path}' terminated!")
          Behaviors.same
      }
    }
}

File: akka-samples/java-samples/src/main/java/actors/watch/WatchPrinter.java
package actors.watch;

import akka.actor.typed.ActorSystem;
import akka.actor.typed.Behavior;
import akka.actor.typed.Terminated;
import akka.actor.typed.javadsl.AbstractBehavior;
import akka.actor.typed.javadsl.ActorContext;
import akka.actor.typed.javadsl.Behaviors;
import akka.actor.typed.javadsl.Receive;

class WatchPrinter extends AbstractBehavior<String> {

    @Override
    public Receive<String> createReceive() {
        return newReceiveBuilder().onMessage(String.class, (message) -> {
            System.out.println(message);
            return Behaviors.same();
        }).onSignal(Terminated.class, (ref) -> {
            System.out.println("Done!");
            return Behaviors.same();
        }).build();
    }

    public WatchPrinter(ActorContext<String> context) {
        super(context);

        getContext().watch(getContext().spawnAnonymous(Worker.create()));
        getContext().watchWith(getContext().spawnAnonymous(Worker.create()), "Worker finished!");
    }

    public static Behavior<String> create() {
        return Behaviors.setup(WatchPrinter::new);
    }

    public static void main(String[] args) {
        ActorSystem.create(WatchPrinter.create(), "watch-printer");
    }
}